### DISCUSSION TITLE
Detect user avatar position in JavaScript

### DISCUSSION THREAD

**Discussion by MarcusLongmuir** [mml-io/mml#29] [Ideas]
### What is your idea?

The ability for JavaScript in the MML document to be able to detect a user's avatar position.

### Why is it important?

This would allow building documents such as pets that follow users around, or have users trigger behaviour simply by them moving into certain regions.

### What do you propose?

Introduce a `m-position-probe` that allows specifying a range attribute within which a user is considered to have entered and then have the user send position updates at a customisable interval. Conceptually similar to `mouseenter`, `mousemove` and `mouseleave` events in regular 2D HTML.

**Comment by docherty** [comment]
I view this similar to the trigger devices in UEFN/Fortnite creative. You might get some ideas from here: https://dev.epicgames.com/documentation/en-us/uefn/verse-api/fortnitedotcom/devices/trigger_device

Having made the flying carpet, the interaction in construct is a real problem. This object might solve it if I can create zones onto which the user walks to direct the carpet. Clearly with more than one user on board that would fall apart but if I can send/read the ID of the triggering user (from the event) then I could filter based on a specific userID.

**Comment by michavie** [comment]
This player position-probe concept could offer far more than just novelty; recognizing a player's position forms one of two critical pillars - the other being Identity - that will provide a competitive edge to countless projects through two key areas: Personalization and Automation

Think about this: no IKEA in the real world can modify its store layout to cater precisely to your preferences and needs, but in the metaverse, it can. A position probe, coupled with Identity, becomes vital for this

Of course, that's just one example in a sea of potential applications. Imagine the enhancements to player experiences and immersion - automated attendance proofs, checkpoint logs, access control to specified zones, or even simple conveniences like lights automatically turning on when you enter an area at night.

Furthermore, with advent of potential widgets, this could lead to adaptive widgets to an extend even your iphone can't be customized. Like a scoreboard popping up when you leave an (e)sports arena, to keep following game highlights

Really just trying to spark ideas here

To summarize, I am all for the addition of an `<m-position-probe>` to MML. The potential use cases, especially when paired with Identity, enable exceptional competitive advantages to real-world businesses that choose the Metaverse

**Comment by prankcall** [comment]
Companion animals are a point of emphasis when my kids play Minecraft, will play a critical role as AI-NPCs play increasing roles, and this type of function is an important construct in mini-game design. 

For this addition.

**Comment by MarcusLongmuir** [comment]
The user position prototype in the form of `m-position-probe` is now in an experimental branch on the MML Playground: https://github.com/mml-io/mml-playground/tree/experiment/m-position-probe. If any intrepid MML explorer wants to give it an early test you'll need to check out the MML repo too (same branch name): https://github.com/mml-io/mml/tree/experiment/m-position-probe and then:
```bash
npm install # in both repos
npm run link-all # in the mml repo
npm link mml-web # in the mml-playground repo
npm run iterate # in both repos.
```
Then go to http://localhost:8080 to access the MML Playground running on your machine

**Comment by MarcusLongmuir** [comment]
This is now tracked by #64.

**Comment by MarcusLongmuir** [comment]
This has been released in https://github.com/mml-io/mml/releases/tag/v0.2.0 🎉
